ports/raspberrypi/common-hal/pulseio/PulseIn.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Dave Putz for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "src/rp2_common/hardware_gpio/include/hardware/gpio.h"
+#include "src/rp2_common/hardware_irq/include/hardware/irq.h"
+
+#include <stdint.h>
+
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/pulseio/PulseIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "supervisor/shared/translate.h"
+#include "bindings/rp2pio/StateMachine.h"
+#include "common-hal/pulseio/PulseIn.h"
+
+pulseio_pulsein_obj_t* save_self;
+
+#define NO_PIN 0xff
+volatile bool last_level;
+volatile uint16_t level_count = 0;
+volatile uint16_t result = 0;
+volatile uint16_t buf_index = 0;
+
+uint16_t pulsein_program[] = {
+    0x2020, //  0: wait   0 pin, 0
+    0xe03f, //  1: set    x, 31
+    0x4001, //  2: in     pins, 1
+    0x0042, //  3: jmp    x--, 2
+    0x8060, //  4: push   iffull block
+    0xc020, //  5: irq    wait 0
+    0x0001, //  6: jmp    1
+};
+
+void common_hal_pulseio_pulsein_construct(pulseio_pulsein_obj_t* self,
+        const mcu_pin_obj_t* pin, uint16_t maxlen, bool idle_state) {
+
+    self->buffer = (uint16_t *) m_malloc(maxlen * sizeof(uint16_t), false);
+    if (self->buffer == NULL) {
+        mp_raise_msg_varg(&mp_type_MemoryError, translate("Failed to allocate RX buffer of %d bytes"), maxlen * sizeof(uint16_t));
+    }
+    self->pin = pin->number;
+    self->maxlen = maxlen;
+    self->idle_state = idle_state;
+    self->start = 0;
+    self->len = 0;
+    save_self = self;
+    // change initial state machine wait if idle_state is false
+    if (idle_state == false) {
+       pulsein_program[0] = 0x20a0;
```

Comment:
There is a small bug caused by this. The StateMachine code will try to share program memory if it detects that the same program is used twice. It does this based on the program pointer. So, multiple pulseins with different initial states may not work concurrently. The simplest fix is to have two copies of the program here.

Reply:
Reading down, do you still need this? I see you exec the wait below.

Reply:
Yes, we still need this. The exec of the wait is only done during a resume(). Some devices (i.e. irremote) don't do a resume - they just wait for the next pulse to come in. 
       But if we have more than one pulsein device; each one will create its own state machine - won't that then create a new offset (program pointer) for each one? The programs themselves would be different due to the first operator differing if the initial states differ.

Reply:
I think you can do the same exec in construct that you do in resume.

No, state machines will share program offsets if the pointer is the same as before. It probably should xor the program itself but it doesn't now.

Reply:
OK. Moved the exec of the wait state from the program to the construct routine.